### PR TITLE
1brcgo submission - nathan #4

### PR DESCRIPTION
**Pull request by nvb**

### CODE REVIEW

**Comment by nvb** [comment]
```
➜ hyperfine ./nathan
Benchmark 1: ./nathan
  Time (mean ± σ):      1.659 s ±  0.006 s    [User: 14.050 s, System: 1.369 s]
  Range (min … max):    1.652 s …  1.672 s    10 runs
```

File: nathan.go [hunk 1]
```diff
@@ -0,0 +1,1548 @@
...
+	// NOTE: this performs better than `return *unsafeIdxPtr[T](dataPtr, i)`.
+	var v T
+	return *(*T)(unsafe.Add(dataPtr, int(unsafe.Sizeof(v))*i))
+}
+
+// unsafeIdxPtr returns &data[i] without bounds checking.
+func unsafeIdxPtr[T any](dataPtr unsafe.Pointer, i int) *T {
+	var v T
+	return (*T)(unsafe.Add(dataPtr, int(unsafe.Sizeof(v))*i))
+}
```
Comment (pav-kv) [inline nathan.go:49]: Nice, I didn't find a way to do this.

File: nathan.go [hunk 2]
```diff
@@ -0,0 +1,1548 @@
...
+
+// unsafeString returns string(data[i:j]) without bounds checking or copying.
+func unsafeString(dataPtr unsafe.Pointer, i, j int) string {
+	return unsafe.String((*byte)(unsafe.Add(dataPtr, i)), j-i)
+}
+
+// boolToInt converts a bool to an int. It is optimized by the go compiler to
+// avoid branching.
```
Comment (pav-kv) [inline nathan.go:57]: Nice. Did you see that in assembly?

File: nathan.go [hunk 3]
```diff
@@ -0,0 +1,1548 @@
...
+}
+
+// simpleMapSlots is the number of slots in a simpleMap. It is a power of two
+// to allow for fast modulus operations.
+const simpleMapSlots = 1 << 16 // 64K slots
+
+func newSimpleMap[K ~uint64, V any]() simpleMap[K, V] {
+	slots := make([]simpleMapSlot[K, V], simpleMapSlots)
+	return simpleMap[K, V]{
+		ptr: unsafe.Pointer(&slots[0]),
```
Comment (pav-kv) [inline nathan.go:101]: Does `&[0]` do a bound check, or the compiler optimizes it out?

File: nathan.go (hunk 2 above)
Comment (mgartner) [reply nathan.go:57]: I stumbled upon something similar: https://github.com/dhartunian/1brcgo/blob/main/marcus.go#L321-L324

**Pull request merged**